samples/drivers/adc/src/main.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020 Libre Solar Technologies GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/printk.h>
+#include <drivers/adc.h>
+
+#if !DT_NODE_EXISTS(DT_PATH(zephyr_user)) || \
+	!DT_NODE_HAS_PROP(DT_PATH(zephyr_user), io_channels)
+#error "No suitable devicetree overlay specified"
+#endif
+
+#define ADC_NUM_CHANNELS	DT_PROP_LEN(DT_PATH(zephyr_user), io_channels)
+
+#if ADC_NUM_CHANNELS > 2
+#error "Currently only 1 or 2 channels supported in this sample"
+#endif
+
+/* Check that the same ADC is used for first and second phandle */
+#if ADC_NUM_CHANNELS == 2 && \
+	DT_PHANDLE_BY_IDX(DT_PATH(zephyr_user), io_channels, 0) != \
+	DT_PHANDLE_BY_IDX(DT_PATH(zephyr_user), io_channels, 1)
+#error "Channels have to use the same ADC."
+#endif
+
+#define ADC_NODE		DT_PHANDLE(DT_PATH(zephyr_user), io_channels)
+
+/* Check if the specified ADC is enabled */
+#if DT_NODE_HAS_STATUS(ADC_NODE, okay)
+#define ADC_DEVICE_NAME		DT_LABEL(ADC_NODE)
+#else
+#error "ADC disabled or not available for this board"
+#endif
+
+/* Common settings supported by most ADCs */
+#define ADC_RESOLUTION		12
+#define ADC_GAIN		ADC_GAIN_1
+#define ADC_REFERENCE		ADC_REF_INTERNAL
+#define ADC_ACQUISITION_TIME	ADC_ACQ_TIME_DEFAULT
+
+/* Get the numbers of up to two channels */
+static uint8_t channel_ids[ADC_NUM_CHANNELS] = {
+	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 0),
+#if ADC_NUM_CHANNELS == 2
+	DT_IO_CHANNELS_INPUT_BY_IDX(DT_PATH(zephyr_user), 1)
+#endif
+};
+
+static int16_t sample_buffer[ADC_NUM_CHANNELS];
+
+struct adc_channel_cfg channel_cfg = {
+	.gain = ADC_GAIN,
+	.reference = ADC_REFERENCE,
+	.acquisition_time = ADC_ACQUISITION_TIME,
+	/* channel ID will be overwritten below */
+	.channel_id = 0,
+	.differential = 0
+};
+
+struct adc_sequence sequence = {
+	/* individual channels will be added below */
+	.channels    = 0,
+	.buffer      = sample_buffer,
+	/* buffer size in bytes, not number of samples */
+	.buffer_size = sizeof(sample_buffer),
+	.resolution  = ADC_RESOLUTION,
+};
+
+void main(void)
+{
+	int err;
+	const struct device *dev_adc = device_get_binding(ADC_DEVICE_NAME);
+
+	if (!dev_adc) {
+		printk("ADC device not found\n");
+		return;
+	}
+
+	/*
+	 * Configure channels individually prior to sampling
+	 */
+	for (uint8_t i = 0; i < ADC_NUM_CHANNELS; i++) {
+		channel_cfg.channel_id = channel_ids[i];
+		adc_channel_setup(dev_adc, &channel_cfg);
+
+		sequence.channels |= BIT(channel_ids[i]);
+	}
+
+	while (1) {
+		/*
+		 * Read sequence of channels (fails if not supported by MCU)
+		 */
+		err = adc_read(dev_adc, &sequence);
+		if (err != 0) {
+			printk("ADC reading failed with error %d.\n", err);
+			return;
+		}
+
+		printk("Raw ADC reading(s):");
+		for (uint8_t i = 0; i < ADC_NUM_CHANNELS; i++) {
+			printk(" %d", sample_buffer[i]);
```

Comment:
There is API (`adc_raw_to_millivolts()`) that converts the raw reading to millivolts; it would be nice to include that in the sample, as many people will find that more useful than the raw value.

Reply:
Unfortunately, the API `adc_ref_internal()` is not supported by all drivers. Most STM32 MCUs for example use the voltage at an external VDDA pin (which is usually equal to the MCU supply voltage) as the reference and `adc_ref_internal()` always returns 0 (not set by the driver because the supply voltage is unknown).

Without knowing the ADC reference voltage, also the `adc_raw_to_millivolts` API cannot be used properly, so I suggest to keep the raw readings in the sample until handling of the reference voltage is improved.

Reply:
OK.  But you could fetch the reference voltage, and if it's not zero provide the millivolt conversion.  Then the sample would work as desired on boards that do support this feature, and it would be a clue to users for what they need to do to get the measurement in a useful unit.